src/Agent.Worker/TestResults/TestRunPublisher.cs

```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Client;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    [ServiceLocator(Default = typeof(TestRunPublisher))]
+    public interface ITestRunPublisher : IAgentService
+    {
+        void StartTestRun(TestRunData testRunData);
+        void AddResults(TestCaseResultData[] testResults);
+        TestRunData ReadResultsFromFile(string filePath);
+        string EndTestRun(bool publishAttachmentsAsArchive = false);
+        void InitializePublisher(IExecutionContext executionContext, VssConnection connection, string projectName, TestRunContext runContext, IResultReader resultReader);
+        void InitializePublisher(IExecutionContext executionContext, ITestManagementHttpClient httpClient, string projectName, TestRunContext runContext, IResultReader resultReader);
+        TestRunData ReadResultsFromFile(string filePath, string runName);
+    }
+
+    public class TestRunPublisher : AgentService, ITestRunPublisher
+    {
+        #region Private
+        const int BATCH_SIZE = 1000;
+        const int PUBLISH_TIMEOUT = 300;
+        const int TCM_MAX_FILESIZE = 104857600;
+        private IExecutionContext _executionContext;
+        private string _projectName;
+        private ITestManagementHttpClient _httpClient;
+        private TestRun _testRun;
+        private TestRunData _testRunData;
+        private IResultReader _resultReader;
+        private TestRunContext _runContext;
+        private IZipper _zipper;
+        #endregion
+
+        #region Public API
+        public void InitializePublisher(IExecutionContext executionContext, VssConnection connection, string projectName, TestRunContext runContext, IResultReader resultReader)
+        {
+            _executionContext = executionContext;
+            _projectName = projectName;
+            _runContext = runContext;
+            _resultReader = resultReader;
+            connection.InnerHandler.Settings.SendTimeout = TimeSpan.FromSeconds(PUBLISH_TIMEOUT);
+            _httpClient = connection.GetClient<TestManagementHttpClient>();
```

Comment:
better follow the pattern Build and RM had, move all client behind interface, reference to BuildServer.cs and ReleaseServer.cs code.